syntax = "proto3";

package producto;

import "tienda.proto";

service Producto {
  rpc AgregarProducto(AgregarProductoRequest) returns (AgregarProductoResponse) {}
  rpc AgregarTalle(AgregarTalleRequest) returns (AgregarTalleResponse) {}
  rpc AgregarStock(AgregarStockRequest) returns (AgregarStockResponse) {}
  rpc ObtenerProducto(ObtenerProductoRequest) returns (ObtenerProductoResponse) {}
  rpc ModificarProducto(ModificarProductoRequest) returns (ModificarProductoResponse) {}
  rpc EliminarProducto(EliminarProductoRequest) returns (ModificarProductoResponse) {}
  rpc TraerTodosLosProductos(TraerTodosLosProductosRequest) returns (TraerTodosLosProductosResponse) {}
  rpc TraerTodosLosProductosFiltrados(TraerTodosLosProductosFiltradosRequest) returns (TraerTodosLosProductosFiltradosResponse) {}
  rpc TraerProductosXTienda(TraerProductosXTiendaRequest) returns (TraerProductosXTiendaResponse) {}
  rpc TraerProductosFiltradosXTienda(TraerProductosFiltradosXTiendaRequest) returns (TraerProductosFiltradosXTiendaResponse) {}
}

message ProductoGrpcDTO {
  int64 idProducto = 1;
  string nombre = 2;
  string foto = 3;
  string color = 4;
  string codigo = 5;
  bool habilitado = 6;
  string talle = 7;
}

message StockGrpcDTO {
  int64 idProducto = 1;
  string nombre = 2;
  string foto = 3;
  string color = 4;
  string codigo = 5;
  int64 cantidad = 6;
  string talle = 7;
  int64 idStock = 9;
}

message AgregarProductoRequest {
  ProductoGrpcDTO productoGrpcDTO = 1;
  repeated int64 tiendas = 2;
}

message AgregarProductoResponse {
  int64 idProducto = 1;
}

message AgregarTalleRequest {
  int64 idProducto = 1;
  repeated int64 tiendas = 2;
  string talle = 3;
}

message AgregarTalleResponse {
  int64 idStock = 1;
}

message ProductoList {
  repeated ProductoGrpcDTO productos = 1;
}

message StockList {
  repeated StockGrpcDTO productos = 1;
}

message ObtenerProductoRequest {
  int64 idProducto = 1;
  string talle = 2;
}

message ObtenerProductoResponse {
  ProductoGrpcDTO productoGrpcDTO = 1;
  tienda.TiendaList tiendas = 2;
}

message ModificarProductoRequest {
  ProductoGrpcDTO productoGrpcDTO = 1;
  repeated ListaEstado tiendas = 2;
}

message ListaEstado {
  int64 id = 1;  
  bool estado = 2;
}

message ModificarProductoResponse {
  int64 idProducto = 1;
}

message EliminarProductoRequest {
  int64 idProducto = 1;
}

message EliminarProductoResponse {
  int64 idProducto = 1;
}

message TraerTodosLosProductosRequest {
}

message TraerTodosLosProductosResponse {
  ProductoList productoList = 1;
}

message TraerTodosLosProductosFiltradosRequest {
  string nombre = 1;
  string codigo = 2;
  string talle = 3;
  string color = 4;
}

message TraerTodosLosProductosFiltradosResponse {
  ProductoList productoList = 1;
}

message TraerProductosXTiendaRequest {
  int64 idTienda = 1;
}

message TraerProductosXTiendaResponse {
  StockList productoList = 1;
}

message TraerProductosFiltradosXTiendaRequest {
  int64 idTienda = 1;
  string nombre = 2;
  string codigo = 3;
  string talle = 4;
  string color = 5;
}

message TraerProductosFiltradosXTiendaResponse {
  StockList productoList = 1;
}

message AgregarStockRequest {
  int64 idStock = 1;
  int64 cantidad = 2;
}

message AgregarStockResponse {
}